### ISSUE TITLE
Any-resolution Training for High-resolution Image Synthesis

### ISSUE THREAD

**Issue by chaerinkong** [reyllama/paper-reviews#190]
# TL; DR

- Additionally condition StyleGAN3 generator on implicit scale and patch location offset to train it to synthesize realistic image patches of arbitrary resolution.

# Motivation

- Training of generative models are often inefficient as we adopt fixed resolution dataset setting, under which we discard lower resolution samples and downsample higher resolution samples (also loss of information).

# Formulation

![image](https://user-images.githubusercontent.com/46308407/203517682-abfe5c11-a169-40be-8c18-8cf8d2dc8c00.png)

- StyleGAN3 generator takes normalized coordinate grid as input, with mapped latent codes for style modulation.
- For any-resolution training, sample implicit image scale s and patch center location v. Note that the generator operates on fixed patch size of p x p, so **s** has to be greater than or equal to **p**. This is the *implicit* full image resolution.

![image](https://user-images.githubusercontent.com/46308407/203518264-89eafcef-69ee-4de9-b0b9-3d3186249035.png)

- Fourier embed coordinate grid, and apply a mapping network (same architecture as the latent codes) to modulate scale. This is to tell the model which portion of the full image it is currently generating.

### Two-stage Training

- First train the model with standard GAN training, i.e., with fixed-resolution dataset. This corresponds to fixing s=p and v=(0.5, 0.5).

![image](https://user-images.githubusercontent.com/46308407/203518582-f51122fe-5c27-4ea2-8e61-c32a20b84dba.png)

- Then, use this weight as the frozen teacher. In the second phase, any-resolution training is on with 50% probability. (still alternate between fixed-res training. *it must have been quite unstable*)

![image](https://user-images.githubusercontent.com/46308407/203518819-82927d74-774d-40cd-834f-f3764aea5060.png)
![image](https://user-images.githubusercontent.com/46308407/203518860-8d99fbe2-a578-4319-b256-49e9144a1610.png)

- Full objective for the second phase. Where *d* is the sum of pixel l1 loss and LPIPS. 

# Evaluation

- New metric to evaluate realistic texture synthesis capacity: patch-FID.
- As opposed to classic FID that first downsamples all images to 299 (for Inception-v3), this uses random crops (patches) of both the real and generated samples. It can therefore capture high frequency details.

![image](https://user-images.githubusercontent.com/46308407/203519331-33f6011c-5cb0-466b-9f1b-17a6bc9f2884.png)

- It shows that any-resolution training improves training efficiency, as only a small portion of HR image suffice for high-res image synthesis.

![image](https://user-images.githubusercontent.com/46308407/203519445-45b8da12-0526-4c70-b5d0-51a9f7bbb0af.png)
![image](https://user-images.githubusercontent.com/46308407/203519467-233035f9-fa29-49a9-b6f6-a80a77275b39.png)
![image](https://user-images.githubusercontent.com/46308407/203519489-92285780-bdfa-4264-ae6b-4edb72d9901f.png)
![image](https://user-images.githubusercontent.com/46308407/203519506-14629d5d-32a6-4305-9918-72d2af69a29c.png)
![image](https://user-images.githubusercontent.com/46308407/203519549-ec8a1bbe-0391-491a-a66e-c5a2eec853c4.png)
![image](https://user-images.githubusercontent.com/46308407/203519577-b3a208c9-5a72-4be3-a134-8498e64d110e.png)

# Inspiration

- Idea from coordinate-MLP and implicit neural representations, successfully implemented to the state-of-the-art GAN to yield pleasant result.
- The way they argued the need for this setting (throwing away low-res training images, downsampling high-res images) was nice.
- How they presented training efficiency (majority of low res + little high res) as one of the main advantages was interesting.
- Can efficiently scale up previous approaches like INR-GAN, which modeled whole image for every training iteration.
- The key idea maybe is to train with image patches (subsets) to render the framework compute-friendly. Inherently similar to lazy regularizations.
